Incorrect interval math on dates in DST

This issue is present on today's build of CockroachDB master that is NOT present in v20.1.0-alpha.20191216 (December 16 build):

When no timezone is active:
```sql
> CREATE TABLE "example2" ("id" integer NOT NULL PRIMARY KEY DEFAULT unique_rowid(), "a" date NOT NULL, "b" date NOT NULL);
> INSERT INTO "example2" ("a", "b") VALUES ('2011-03-13'::date, '2011-03-13'::date);
> INSERT INTO "example2" ("a", "b") VALUES ('2011-03-14'::date, '2011-03-14'::date);
> SELECT * FROM "example2" WHERE "a" < ("b" + '0 days 0.000000 seconds'::interval);
  id | a | b
-----+---+----
(0 rows)


```
But when a timezone is active, any dates after daylight saving time (March 13, 2011) are returned (which is unexpected):
```sql
> SET TIME ZONE 'America/Chicago';
> SELECT * FROM "example2" WHERE "a" < ("b" + '0 days 0.000000 seconds'::interval);
          id         |             a             |             b
---------------------+---------------------------+----------------------------
  525452323391143937 | 2011-03-14 00:00:00+00:00 | 2011-03-14 00:00:00+00:00
```

The relevant Django test:

```
======================================================================
FAIL: test_date_comparison (expressions.tests.FTimeDeltaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/expressions/tests.py", line 1334, in test_date_comparison
    self.assertEqual(test_set, self.expnames[:i])
AssertionError: Lists differ: ['e0'] != []

First list contains 1 additional elements.
First extra element 0:
'e0'

- ['e0']
+ []
```

--- a/pkg/sql/logictest/testdata/logic_test/timestamp
+++ b/pkg/sql/logictest/testdata/logic_test/timestamp
@@ -255,3 +255,43 @@ query B
 select localtimestamp(3) - localtimestamp <= '1ms'::interval
 ----
 true
+
+# When doing daylight savings comparisons, ensure they compare correctly.
+# Test day before and after DST.
+subtest regression_django-cockroachdb_120
+
+statement ok
+SET TIME ZONE 'America/Chicago'
+
+query B
+SELECT '2011-03-13'::date = '2011-03-13'::timestamp
+----
+true
+
+query B
+SELECT '2011-03-13'::date = '2011-03-13'::timestamptz
+----
+true
+
+query B
+SELECT '2011-03-13'::timestamp = '2011-03-13'::timestamptz
+----
+true
+
+query B
+SELECT '2011-03-14'::date = '2011-03-14'::timestamp
+----
+true
+
+query B
+SELECT '2011-03-14'::date = '2011-03-14'::timestamptz
+----
+true
+
+query B
+SELECT '2011-03-14'::timestamp = '2011-03-14'::timestamptz
+----
+true
+
+statement ok
+SET TIME ZONE 0

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -2163,12 +2163,13 @@ func timeFromDatumForComparison(ctx *EvalContext, d Datum) (time.Time, bool) {
 		return t.Time, true
 	case *DTimestamp:
 		// Normalize to the timezone of the context.
-		_, zoneOffset := ctx.GetRelativeParseTime().Zone()
-		return t.Time.In(ctx.GetLocation()).Add(-time.Duration(zoneOffset) * time.Second), true
+		_, zoneOffset := t.Time.In(ctx.GetLocation()).Zone()
+		ts := t.Time.In(ctx.GetLocation()).Add(-time.Duration(zoneOffset) * time.Second)
+		return ts, true
 	case *DTime:
 		// Normalize to the timezone of the context.
 		toTime := timeofday.TimeOfDay(*t).ToTime()
-		_, zoneOffsetSecs := ctx.GetRelativeParseTime().Zone()
+		_, zoneOffsetSecs := toTime.In(ctx.GetLocation()).Zone()
 		return toTime.In(ctx.GetLocation()).Add(-time.Duration(zoneOffsetSecs) * time.Second), true
 	case *DTimeTZ:
 		return t.ToTime(), true

--- a/pkg/util/timetz/timetz.go
+++ b/pkg/util/timetz/timetz.go
@@ -62,7 +62,7 @@ func MakeTimeTZ(t timeofday.TimeOfDay, offsetSecs int32) TimeTZ {
 
 // MakeTimeTZFromLocation creates a TimeTZ from a TimeOfDay and time.Location.
 func MakeTimeTZFromLocation(t timeofday.TimeOfDay, loc *time.Location) TimeTZ {
-	_, zoneOffsetSecs := timeutil.Unix(0, 0).In(loc).Zone()
+	_, zoneOffsetSecs := timeutil.Now().In(loc).Zone()
 	return TimeTZ{TimeOfDay: t, OffsetSecs: -int32(zoneOffsetSecs)}
 }
 

